File: backend/src/main/java/org/luxons/sevenwonders/game/Game.java
package org.luxons.sevenwonders.game;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.luxons.sevenwonders.game.api.Action;
import org.luxons.sevenwonders.game.api.HandCard;
import org.luxons.sevenwonders.game.api.PlayerMove;
import org.luxons.sevenwonders.game.api.PlayerTurnInfo;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.cards.Card;
import org.luxons.sevenwonders.game.cards.CardBack;
import org.luxons.sevenwonders.game.cards.Decks;
import org.luxons.sevenwonders.game.cards.Hands;
import org.luxons.sevenwonders.game.effects.SpecialAbility;
import org.luxons.sevenwonders.game.moves.InvalidMoveException;
import org.luxons.sevenwonders.game.moves.Move;
import org.luxons.sevenwonders.game.scoring.ScoreBoard;

public class Game {

    private static final int LAST_AGE = 3;

    private final long id;

    private final Settings settings;

    private final int nbPlayers;

    private final Table table;

    private final Decks decks;

    private final List<Card> discardedCards;

    private final Map<Integer, Move> preparedMoves;

    private Map<Integer, PlayerTurnInfo> currentTurnInfo;

    private Hands hands;

    public Game(long id, Settings settings, int nbPlayers, List<Board> boards, Decks decks) {
        this.id = id;
        this.settings = settings;
        this.nbPlayers = nbPlayers;
        this.table = new Table(boards);
        this.decks = decks;
        this.discardedCards = new ArrayList<>();
        this.currentTurnInfo = new HashMap<>();
        this.preparedMoves = new HashMap<>();
        startNewAge();
    }

    public long getId() {
        return id;
    }

    public Settings getSettings() {
        return settings;
    }

    private void startNewAge() {
        table.increaseCurrentAge();
        hands = decks.deal(table.getCurrentAge(), table.getNbPlayers());
        startNewTurn();
    }

    private void startNewTurn() {
        currentTurnInfo.clear();
        for (int i = 0; i < nbPlayers; i++) {
            currentTurnInfo.put(i, createPlayerTurnInfo(i));
        }
    }

    private PlayerTurnInfo createPlayerTurnInfo(int playerIndex) {
        PlayerTurnInfo pti = new PlayerTurnInfo(playerIndex, table);
        List<HandCard> hand = hands.createHand(table, playerIndex);
        pti.setHand(hand);
        Action action = determineAction(hand, table.getBoard(playerIndex));
        pti.setAction(action);
        pti.setMessage(action.getMessage());
        if (action == Action.PICK_NEIGHBOR_GUILD) {
            pti.setNeighbourGuildCards(table.getNeighbourGuildCards(playerIndex));
        }
        return pti;
    }

    public Collection<PlayerTurnInfo> getCurrentTurnInfo() {
        return currentTurnInfo.values();
    }

    private Action determineAction(List<HandCard> hand, Board board) {
        if (endOfGameReached() && board.hasSpecial(SpecialAbility.COPY_GUILD)) {
            return Action.PICK_NEIGHBOR_GUILD;
        } else if (hand.size() == 1 && board.hasSpecial(SpecialAbility.PLAY_LAST_CARD)) {
            return Action.PLAY_LAST;
        } else if (hand.size() == 2 && board.hasSpecial(SpecialAbility.PLAY_LAST_CARD)) {
            return Action.PLAY_2;
        } else if (hand.isEmpty()) {
            return Action.WAIT;
        } else {
            return Action.PLAY;
        }
    }

    public CardBack prepareMove(int playerIndex, PlayerMove playerMove) throws InvalidMoveException {
        Card card = decks.getCard(playerMove.getCardName());
        Move move = playerMove.getType().resolve(playerIndex, card, playerMove);
        validate(move);
        preparedMoves.put(playerIndex, move);
        return card.getBack();
    }

    private void validate(Move move) throws InvalidMoveException {
        List<Card> hand = hands.get(move.getPlayerIndex());
        move.validate(table, hand);
    }

    public boolean areAllPlayersReady() {
        long nbExpectedMoves = currentTurnInfo.values().stream().filter(pti -> pti.getAction() != Action.WAIT).count();
        return preparedMoves.size() == nbExpectedMoves;
    }

    public void playTurn() {
        makeMoves();
        if (endOfAgeReached()) {
            executeEndOfAgeEvents();
            if (!endOfGameReached()) {
                startNewAge();
            }
        } else {
            rotateHandsIfRelevant();
            startNewTurn();
        }
    }

    private void rotateHandsIfRelevant() {
        // we don't rotate hands if some player can play his last card (with the special ability)
        if (!hands.maxOneCardRemains()) {
            hands.rotate(table.getHandRotationDirection());
        }
    }

    private void makeMoves() {
        List<Move> playedMoves = mapToList(preparedMoves);

        // all cards from this turn need to be placed before executing any effect
        // because effects depending on played cards need to take the ones from the current turn into account too
        placePreparedCards(playedMoves);

        // same goes for the discarded cards during the last turn, which should be available for special actions
        if (hands.maxOneCardRemains()) {
            discardLastCardsOfHands();
        }

        activatePlayedCards(playedMoves);

        table.setLastPlayedMoves(playedMoves);
        preparedMoves.clear();
    }

    private static List<Move> mapToList(Map<Integer, Move> movesPerPlayer) {
        List<Move> moves = new ArrayList<>(movesPerPlayer.size());
        for (int p = 0; p < movesPerPlayer.size(); p++) {
            Move move = movesPerPlayer.get(p);
            if (move == null) {
                throw new MissingPreparedMoveException(p);
            }
            moves.add(move);
        }
        return moves;
    }

    private void placePreparedCards(List<Move> playedMoves) {
        playedMoves.forEach(move -> {
            move.place(table, discardedCards, settings);
            removeFromHand(move.getPlayerIndex(), move.getCard());
        });
    }

    private void discardLastCardsOfHands() {
        for (int i = 0; i < nbPlayers; i++) {
            Board board = table.getBoard(i);
            if (!board.hasSpecial(SpecialAbility.PLAY_LAST_CARD)) {
                discardHand(i);
            }
        }
    }

    private void discardHand(int playerIndex) {
        List<Card> hand = hands.get(playerIndex);
        discardedCards.addAll(hand);
        hand.clear();
    }

    private void removeFromHand(int playerIndex, Card card) {
        hands.get(playerIndex).remove(card);
    }

    private void activatePlayedCards(List<Move> playedMoves) {
        playedMoves.forEach(move -> move.activate(table, discardedCards, settings));
    }

    private boolean endOfAgeReached() {
        return hands.isEmpty();
    }

    private void executeEndOfAgeEvents() {
        table.resolveMilitaryConflicts();
    }

    private boolean endOfGameReached() {
        return endOfAgeReached() && table.getCurrentAge() == LAST_AGE;
    }

    public ScoreBoard computeScore() {
        ScoreBoard scoreBoard = new ScoreBoard();
        table.getBoards().stream().map(b -> b.computePoints(table)).forEach(scoreBoard::add);
        return scoreBoard;
    }

    private static class MissingPreparedMoveException extends IllegalStateException {
        MissingPreparedMoveException(int playerIndex) {
            super("Player " + playerIndex + " is not ready to play");
        }
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/Game.kt
package org.luxons.sevenwonders.game

import org.luxons.sevenwonders.game.api.Action
import org.luxons.sevenwonders.game.api.HandCard
import org.luxons.sevenwonders.game.api.PlayerMove
import org.luxons.sevenwonders.game.api.PlayerTurnInfo
import org.luxons.sevenwonders.game.api.toApiTable
import org.luxons.sevenwonders.game.api.toTableCard
import org.luxons.sevenwonders.game.boards.Board
import org.luxons.sevenwonders.game.boards.Table
import org.luxons.sevenwonders.game.cards.Card
import org.luxons.sevenwonders.game.cards.CardBack
import org.luxons.sevenwonders.game.cards.Decks
import org.luxons.sevenwonders.game.cards.Hands
import org.luxons.sevenwonders.game.data.LAST_AGE
import org.luxons.sevenwonders.game.effects.SpecialAbility
import org.luxons.sevenwonders.game.moves.Move
import org.luxons.sevenwonders.game.score.ScoreBoard
import org.luxons.sevenwonders.game.api.Table as ApiTable

class Game internal constructor(
    val id: Long,
    private val settings: Settings,
    boards: List<Board>,
    private val decks: Decks
) {
    private val table: Table = Table(boards)
    private val players: List<Player> = boards.map { SimplePlayer(it.playerIndex, table) }
    private val discardedCards: MutableList<Card> = mutableListOf()
    private val preparedMoves: MutableMap<Int, Move> = mutableMapOf()
    private var currentTurnInfo: List<PlayerTurnInfo> = emptyList()
    private var hands: Hands = Hands(emptyList())

    init {
        startNewAge()
    }

    private fun startNewAge() {
        table.increaseCurrentAge()
        hands = decks.deal(table.currentAge, players.size)
        startNewTurn()
    }

    private fun startNewTurn() {
        currentTurnInfo = players.map { createPlayerTurnInfo(it) }
    }

    private fun createPlayerTurnInfo(player: Player): PlayerTurnInfo {
        val hand = hands.createHand(player)
        val action = determineAction(hand, player.board)
        val neighbourGuildCards = table.getNeighbourGuildCards(player.index).map { it.toTableCard(null) }

        return PlayerTurnInfo(player.index, table.toApiTable(), action, hand, neighbourGuildCards)
    }

    /**
     * Returns information for each player about the current turn.
     */
    fun getCurrentTurnInfo(): Collection<PlayerTurnInfo> = currentTurnInfo

    private fun determineAction(hand: List<HandCard>, board: Board): Action = when {
        endOfGameReached() && board.hasSpecial(SpecialAbility.COPY_GUILD) -> determineCopyGuildAction(board)
        hand.size == 1 && board.hasSpecial(SpecialAbility.PLAY_LAST_CARD) -> Action.PLAY_LAST
        hand.size == 2 && board.hasSpecial(SpecialAbility.PLAY_LAST_CARD) -> Action.PLAY_2
        hand.isEmpty() -> Action.WAIT
        else -> Action.PLAY
    }

    private fun determineCopyGuildAction(board: Board): Action {
        val neighbourGuildCards = table.getNeighbourGuildCards(board.playerIndex)
        return if (neighbourGuildCards.isEmpty()) Action.WAIT else Action.PICK_NEIGHBOR_GUILD
    }

    /**
     * Prepares the given [move] for the player at the given [playerIndex].
     *
     * @return the back of the card that is prepared on the table
     */
    fun prepareMove(playerIndex: Int, move: PlayerMove): CardBack {
        val card = decks.getCard(table.currentAge, move.cardName)
        val context = PlayerContext(playerIndex, table, hands[playerIndex])
        val resolvedMove = move.type.resolve(move, card, context)
        preparedMoves[playerIndex] = resolvedMove
        return card.back
    }

    /**
     * Returns true if all players that had to do something have [prepared their move][prepareMove]. This means we are
     * ready to [play the current turn][playTurn].
     */
    fun allPlayersPreparedTheirMove(): Boolean {
        val nbExpectedMoves = currentTurnInfo.count { it.action !== Action.WAIT }
        return preparedMoves.size == nbExpectedMoves
    }

    /**
     * Plays all the [prepared moves][prepareMove] for the current turn. An exception will be thrown if some players
     * had not prepared their moves (unless these players had nothing to do). To avoid this, please check if everyone
     * is ready using [allPlayersPreparedTheirMove].
     */
    fun playTurn(): ApiTable {
        makeMoves()
        if (endOfAgeReached()) {
            executeEndOfAgeEvents()
            if (!endOfGameReached()) {
                startNewAge()
            }
        } else {
            rotateHandsIfRelevant()
            startNewTurn()
        }
        return table.toApiTable()
    }

    private fun makeMoves() {
        val moves = getMovesToPerform()

        // all cards from this turn need to be placed before executing any effect
        // because effects depending on played cards need to take the ones from the current turn into account too
        placePreparedCards(moves)

        // same goes for the discarded cards during the last turn, which should be available for special actions
        if (hands.maxOneCardRemains()) {
            discardLastCardsOfHands()
        }

        activatePlayedCards(moves)

        table.lastPlayedMoves = moves
        preparedMoves.clear()
    }

    private fun getMovesToPerform(): List<Move> =
        currentTurnInfo.filter { it.action !== Action.WAIT }.map { getMoveToPerformFor(it.playerIndex) }

    private fun getMoveToPerformFor(playerIndex: Int) =
        preparedMoves[playerIndex] ?: throw MissingPreparedMoveException(playerIndex)

    private fun endOfAgeReached(): Boolean = hands.isEmpty

    private fun executeEndOfAgeEvents() = table.resolveMilitaryConflicts()

    private fun endOfGameReached(): Boolean = endOfAgeReached() && table.currentAge == LAST_AGE

    private fun rotateHandsIfRelevant() {
        // we don't rotate hands if some player can play his last card (with the special ability)
        if (!hands.maxOneCardRemains()) {
            hands = hands.rotate(table.handRotationDirection)
        }
    }

    private fun placePreparedCards(playedMoves: List<Move>) {
        playedMoves.forEach { move ->
            move.place(discardedCards, settings)
            hands = hands.remove(move.playerContext.index, move.card)
        }
    }

    private fun discardLastCardsOfHands() =
        table.boards.filterNot { it.hasSpecial(SpecialAbility.PLAY_LAST_CARD) }.forEach { discardHand(it.playerIndex) }

    private fun discardHand(playerIndex: Int) {
        val hand = hands[playerIndex]
        discardedCards.addAll(hand)
        hands = hands.discardHand(playerIndex)
    }

    private fun activatePlayedCards(playedMoves: List<Move>) =
        playedMoves.forEach { it.activate(discardedCards, settings) }

    /**
     * Computes the score for all players.
     */
    fun computeScore(): ScoreBoard = ScoreBoard(table.boards.map { it.computeScore(players[it.playerIndex]) })

    private class MissingPreparedMoveException internal constructor(playerIndex: Int) :
        IllegalStateException("Player $playerIndex has not prepared his move")
}
